state/spaces.go

```diff
@@ -267,6 +267,34 @@ func (st *State) SpaceByName(name string) (*Space, error) {
 	return &Space{st, doc}, nil
 }
 
+// SpaceByNameAndID returns a slice of spaces from state that matches the input name or ID.
+// Multiple matches will be returned in case the name and spaceID are matched.
+// An error is returned if the space does not exist or if there was a problem
+// accessing its information.
+func (st *State) SpaceByNameAndID(nameOrID string) ([]*Space, error) {
```

Comment:
Instead of adding this method, just put logic in the API server that queries by name, and only queries by ID if the name it is an integer.

Reply:
I had that initially. But the name itself can also be an integer, right?
If I put the logic into the API server I would need to do 2 queries in the worst case. Worst case being someone calls his space `0`, like the alpha space.
```
if name==int{
space,err = spaceById(name)
if notfound(err) {
space,err = spaceByName(name)
}
}
```

Reply:
Hmm... Note that space names starting with a number are [valid](https://github.com/juju/names/blob/v3/space.go#L13) space names. If we go down this route, we _may_ cause potential issues if the operators opt to use one-digit space names.

For example, if I have a space with name "1", and have more than 2 spaces defined, I might end up with something like this in the db:

```
[
  {spaceid: 1, name: "2"},
  {spaceid: 2, name: "1"}
]
```

If I pass "2" to this method, I will get back 2 results and trigger this error: https://github.com/juju/juju/pull/11195/files#diff-234aab187115a97732fba3ced20d64d6R371-R377

Reply:
Agree, had that discussion with @manadart initially. We came to the conclusion it makes sense to error out and send instructions for the operator to disambiguate e.g
Note: spaceID `0` is the alpha space ID, and a `spaceName` I've created before
```
❯ juju add-space 0

❯ juju show-space 0
{}
ERROR cannot retrieve space "0": multiple spaces found with name/ID "0". Spaces found: ["alpha" "0"]. Please use unique name or ID
```

Reply:
I agree with manadart that this logic should live in the facade and not in state. You can just get `AllSpaces()` (they are not that many anyway) and either define a typed []Space list with filtering methods or filter directly in the facade. 

For example, you can iterate the space list and add name-based matches to one list and ID-based matches to another. Then, you can simply check `len(listA) + len(listB) == 1`.

Reply:
> Agree, had that discussion with @manadart initially. We came to the conclusion it makes sense to error out and send instructions for the operator to disambiguate e.g
> Note: spaceID `0` is the alpha space ID, and a `spaceName` I've created before
> 
> ```
> ❯ juju add-space 0
> 
> ❯ juju show-space 0
> {}
> ERROR cannot retrieve space "0": multiple spaces found with name/ID "0". Spaces found: ["alpha" "0"]. Please use unique name or ID
> ```
> 
> But I agree, this can be totally confusing ...
> 
> If I don't do it this way, it is difficult to know whether the operator wanted with `0` the alpha ID `0` or actually the `spaceName` with `0`.
> 
> We could enforce a hierarchy ordering. If it is an `int` we search for the int first. If we found something we return it. Even if the name was meant.
> 
> More of an UX discussion maybe? @mitechie may want to give input what he prefers.

In my scenario you wouldn't be able to disambiguate because of the way the space ID/names are laid out (probably uncommon to end up with something like that but with the space renaming stuff you never know...)

Reply:
@achilleasa oh, you are right. In this case, I would probably go with the AllSpaces approach. If the number is low this should be fine. 

I will figure out a good example case.